crates/oxc_linter/src/rules/for_direction.rs

```diff
@@ -0,0 +1,156 @@
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint(for-direction): Disallow for loop in the wrong direction")]
+#[diagnostic(severity(warning), help("Use while loop for intended infinite loop"))]
+struct ForDirectionDiagnostic(#[label] pub Span);
+
+#[derive(Debug, Default, Clone)]
+pub struct ForDirection;
+
+declare_oxc_lint!(
+    /// ### What it does
+    /// Disallow "for" loop update causing the counter to move in the wrong direction.
+    ///
+    /// ### Why is this bad?
+    /// A for loop that is known to run infinitely or never run is considered a bug.
+    ///
+    ///
+    /// ### Example
+    /// ```javascript
+    /// for (var i = 0; i < 10; i--) {}
+    ///
+    /// for (var = 10; i >= 0; i++) {}
+    /// ```
+    ForDirection,
+    correctness
+);
+
+type UpdateDirection = i32;
+const FORWARD: UpdateDirection = 1;
+const BACKWARD: UpdateDirection = -1;
+const UNKNOWN: UpdateDirection = 0;
+
+fn get_update_direction(update: &Expression, counter: &IdentifierReference) -> UpdateDirection {
+    match update {
+        // match increment or decrement
+        Expression::UpdateExpression(update) => {
+            if let SimpleAssignmentTarget::AssignmentTargetIdentifier(id) = &update.argument {
+                if id.name != counter.name {
+                    return UNKNOWN;
+                }
+                match update.operator {
+                    UpdateOperator::Increment => FORWARD,
+                    UpdateOperator::Decrement => BACKWARD,
+                }
+            } else {
+                UNKNOWN
+            }
+        }
+        // match add assign or subtract assign
+        Expression::AssignmentExpression(assign) => {
+            if let AssignmentTarget::SimpleAssignmentTarget(
+                SimpleAssignmentTarget::AssignmentTargetIdentifier(id),
+            ) = &assign.left
+            {
+                if id.name != counter.name {
+                    return UNKNOWN;
+                }
+                get_assignment_direction(assign)
+            } else {
+                UNKNOWN
+            }
+        }
+        // can't determine other kinds of updates
+        _ => UNKNOWN,
+    }
+}
+
+fn get_assignment_direction(assign: &AssignmentExpression) -> UpdateDirection {
+    let operator = &assign.operator;
+    let right = &assign.right;
+    let positive = match right {
+        Expression::NumberLiteral(r) => r.value.is_sign_positive(),
+        Expression::UnaryExpression(right) => right.operator != UnaryOperator::UnaryNegation,
+        _ => return UNKNOWN,
+    };
+
+    let mut direction = match operator {
+        AssignmentOperator::Addition => FORWARD,
+        AssignmentOperator::Subtraction => BACKWARD,
+        _ => return UNKNOWN,
+    };
+
+    if !positive {
+        direction = -direction;
+    }
+    direction
+}
+
+impl Rule for ForDirection {
```

Comment:
Let's move this block to the top, just under `declare_oxc_lint` so it's easier for people to read and navigate.

Reply:
Done.